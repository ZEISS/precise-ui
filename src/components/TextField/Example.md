**Elementary**

Various versions of the text field. By default, we get a plain text field that is managed and enabled. No placeholder is shown. The width of the text field is the width of the parent container.

```jsx
const { TextField } = require('precise-ui');

<TextField />
```

**Decoration**

The `label` prop is used to determine what text should be shown as a label.

```jsx
const { TextField } = require('precise-ui');

<TextField label="Label" />
```

The `placeholder` prop is used to determine what text should be shown as a placeholder (for the empty text field).

```jsx
const { TextField } = require('precise-ui');

<TextField label="With label" placeholder="With placeholder" />
```

The `disabled` prop is indicating that the value cannot be edited, because the control is disabled. There is no `readonly` prop, as such a prop is given indirectly via not updating the `value` prop in controlled mode.

```jsx
const { TextField } = require('precise-ui');

<TextField disabled placeholder="disabled"/>
```

The `clearable` prop is showing a little icon to reset the `TextField` as soon as the user enters data.

```jsx
const { TextField } = require('precise-ui');

<TextField clearable placeholder="Enter something ..."/>
```

**Controlled Mode**

The controlled mode is triggered via the `value` prop.

```jsx
const { TextField } = require('precise-ui');

<TextField value="With value" placeholder="With value"/>
```

You can also add a callback to the clear button by using the `onClear` prop.

```jsx
const { TextField } = require('precise-ui');

<TextField value="With value" placeholder="With value" clearable onClear={() => alert('Clear Pressed')}/>
```

This enables us to also fully utilize a controlled mode. Note, that not only the `onClear` is fired when clearing, but also the `onChange` as we want to change to a new value.

```jsx
const { TextField } = require('precise-ui');

class ControlledClearableTextField extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
        value: '',
    };
    this.changed = ({ value }) => {
      this.setState({
          value,
      });
    };
  }

  render() {
      return <TextField clearable value={this.state.value} onChange={this.changed} placeholder="Enter something ..."/>;
}
}

<ControlledClearableTextField />
```

Likewise, we can stay in managed mode by using the `defaultValue` instead.

```jsx
const { TextField } = require('precise-ui');

<TextField defaultValue="With value" placeholder="With default value"/>
```

**Decoration Options**

The text field can be decorated with several options, e.g., adding a `prefix` or `suffix`.

```jsx
const { TextField, Icon } = require('precise-ui');

<div>
  <TextField label="Full Name" prefix={<Icon name="Person" width="100%" height="100%" />} />
  <br />
  <TextField label="E-mail" suffix="@mail.com" />
  <br />
  <TextField label="Website" prefix="https://" suffix=".com" />
</div>
```

We can also indicate errors by placing some text on the `error` prop.

```jsx
const { TextField } = require('precise-ui');

class MyInputForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      value: '',
    };
    this.changeValue = (e) => {
      this.setState({
        value: e.value,
      });
    };
  }

  render() {
    const { value } = this.state;
    const error = value === 'email' ? '' : 'Value must be "email"';
    return <TextField label="Email" defaultValue="emai" error={error} value={value} onChange={this.changeValue} />;
  }
}

<MyInputForm />
```

**Multiline Input**

Finally, multi-line input is also supported. Once we go into multi-line mode the `prefix` and `suffix` props are no longer supported. By default, the multi-line input is not resizable. This can be changed, however, with the `resizable` prop.

```jsx
const { TextField } = require('precise-ui');

<TextField multiline resizable defaultValue="Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Integer pellentesque quam vel velit. Duis pulvinar. Vivamus porttitor turpisacleo. Curabitur bibendum justo non orci. Integer lacinia."/>
```

You can also add the `label` prop to multiline textfields

```jsx
const { TextField } = require('precise-ui');

<TextField multiline resizable label="Test label" />
```

You can also add the `label` prop in addition to the placeholder

```jsx
const { TextField } = require('precise-ui');

<TextField multiline resizable label="Test label" placeholder="Test placeholder" />
```

The resizable can also be restricted to vertical and horizontal.

```jsx
const { TextField } = require('precise-ui');

<TextField multiline resizable="vertical" />
```

The `multiline` prop can be specified as a number, indicating the number of rows initially shown.

```jsx
const { TextField } = require('precise-ui');

<TextField multiline={4} value="My text here" disabled />
```

Last, but not least we can use the `resizable` prop with the `auto` value to automatically adjust the height of the input.

```jsx
const { TextField } = require('precise-ui');

<TextField multiline resizable="auto" defaultValue="My text here" />
```
