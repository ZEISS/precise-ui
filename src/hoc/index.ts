export * from './withResponsive';
export * from './withResponsiveMode';
export * from './withValidator';
export * from './withOutsideClick';
export * from './withFormContext';
