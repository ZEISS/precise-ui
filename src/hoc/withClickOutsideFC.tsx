// This HOC is an adapter allowing functional components to use 'react-clickoutside' package.
// The problem with 'react-clickoutside' is that it only supports class components, and
// the workaround, described in its README does not always work and is ugly.
// This HOC hides imperative nature of 'react-onclickoutside' and notifies the wrapped component
// about outside clicks by passing it the click event as a prop. As each event object is unique,
// the wrapped component may react on clicks with the help of 'useEffect' hook set up to listen
// changes of `props.outsideClickEvent`.

import onClickOutside, { AdditionalProps } from 'react-onclickoutside';
import * as React from 'react';

interface OutsideClickProps {
  outsideClickEvent: React.SyntheticEvent | undefined;
}

export function withClickOutsideFC<P extends object>(Component: React.FC<P>) {
  class OutsideClickAdapter extends React.Component<P, AdditionalProps & OutsideClickProps> {
    static displayName: string;
    constructor(props: P) {
      super(props);
      this.state = {
        outsideClickEvent: undefined,
      };
    }

    handleClickOutside = (e: React.SyntheticEvent) => {
      this.setState({ outsideClickEvent: e });
    };

    render() {
      const { children, ...otherProps } = this.props;
      return (
        <Component outsideClickEvent={this.state.outsideClickEvent} {...otherProps as P}>
          {children}
        </Component>
      );
    }
  }

  OutsideClickAdapter.displayName = 'withClickOutsideWrapper';

  return onClickOutside(OutsideClickAdapter);
}

export interface WithClickOutsideFCProps extends AdditionalProps, OutsideClickProps {}
