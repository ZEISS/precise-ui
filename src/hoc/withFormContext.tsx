import * as React from 'react';
import { InputProps } from '../common';
import { FormContextType, FormContext } from '../contexts/FormContext';

export interface FormContextProps {
  form?: FormContextType;
}

export function withFormContext<T extends InputProps<any>>(
  Component: React.ComponentType<T & FormContextProps>,
): React.SFC<T> {
  return (props: T) =>
    props.name ? (
      <FormContext.Consumer>{ctx => <Component form={ctx} {...props} />}</FormContext.Consumer>
    ) : (
      <Component {...props} />
    );
}
