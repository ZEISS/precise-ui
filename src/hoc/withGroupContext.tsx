import * as React from 'react';
import { RadioButtonGroupContext, RadioButtonGroupContextType } from '../contexts/RadioButtonGroupContext';
import { withInner } from 'typescript-plugin-inner-jsx/withInner';
import { InputProps } from '../common';

export interface GroupContextProps {
  group?: RadioButtonGroupContextType;
}

export function withGroupContext<TProps extends InputProps<any>>(
  Component: React.ComponentType<TProps & GroupContextProps>,
): React.SFC<TProps> {
  return withInner(
    (props: TProps) =>
      props.name ? (
        <RadioButtonGroupContext.Consumer>
          {ctx => <Component group={ctx} {...props} />}
        </RadioButtonGroupContext.Consumer>
      ) : (
        <Component {...props} />
      ),
    { Component },
  );
}
