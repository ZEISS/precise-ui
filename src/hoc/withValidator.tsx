import * as React from 'react';

export interface BaseInputProps<TEventArgs> {
  /**
   * Event emitted once the value changes.
   */
  onChange?(e: TEventArgs): void;
  /**
   * The currently displayed error message.
   */
  error?: React.ReactChild;
}

export interface ValidatorProps {
  onSuccess?(): void;
  onError?(): void;
}

export interface ValidatorState {
  error: React.ReactChild | undefined;
}

export function withValidator<TEventArgs, TProps extends BaseInputProps<TEventArgs>>(
  Component: React.ComponentType<TProps>,
  validate: (e: TEventArgs) => React.ReactChild | undefined,
) {
  return class Response extends React.PureComponent<TProps & ValidatorProps, ValidatorState> {
    constructor(props: TProps & ValidatorProps) {
      super(props);
      this.state = {
        error: undefined,
      };
    }

    validate = (e: TEventArgs) => {
      const { onChange, onError, onSuccess } = this.props;
      const error = validate(e);

      if (error !== this.state.error) {
        const notify = error ? onError : onSuccess;

        this.setState({
          error,
        });

        if (typeof notify === 'function') {
          notify();
        }
      }

      if (typeof onChange === 'function') {
        onChange(e);
      }
    };

    render() {
      const { error } = this.state;
      return <Component {...this.props} error={error || this.props.error} onChange={this.validate} />;
    }
  };
}
