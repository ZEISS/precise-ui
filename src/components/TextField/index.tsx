import * as React from 'react';
import styled, { themed } from '../../utils/styled';
import { StackPanel } from '../StackPanel';
import { purpleRed } from '../../colors';
import { InputIcon } from '../InputIcon';
import { InputChangeEvent, TextInputProps } from '../../common';
import { FormContextProps, withFormContext } from '../../hoc/withFormContext';
import { showInputInfo } from '../../utils/input';
import { distance } from '../../distance';
import { StyledInputRow, StyledInput, TextFieldBorderType, StyledInputBox, getTextFieldBorderType } from '../../quarks';
import { getFontSize } from '../../textStyles';

export type TextFieldChangeEvent = InputChangeEvent<string>;

type TextFieldInputNodeType = HTMLTextAreaElement | HTMLInputElement;

export interface TextFieldProps extends TextInputProps {
  /**
   * Allows multi-line input from the user. Either a boolean
   * or a number specifying the amount of rows to show.
   * @default false
   */
  multiline?: boolean | number;
  /**
   * Allows resizing the text field if being used as a multi-line input.
   * @default false
   */
  resizable?: boolean | 'auto' | 'vertical' | 'horizontal';
  /**
   * Shows a clearing icon which resets the input to an empty string.
   * @default false
   */
  clearable?: boolean;

  /**
   * Event emitted when the clear button was pressed. Will always be fired after
   * the onChange event, i.e., after the value was set / proposed.
   */
  onClear?(): void;

  /**
   * Gets the reference to the underlying input or textarea element.
   * @ignore
   */
  inputRef?(instance: HTMLElement | null): void;

  /**
   * @ignore
   */
  children?: void;
}

export interface TextFieldState {
  controlled: boolean;
  focused: boolean;
  reveal: boolean;
  value: string;
  error?: React.ReactChild;
}

interface TextFieldAreaProps {
  border: TextFieldBorderType;
  disabled?: boolean;
  resizable?: string;
  labelShown: boolean;
}

const TextFieldContainer = styled.div`
  position: relative;
`;

const TextFieldWrapper = styled(StackPanel)`
  display: flex;
`;

const TextFieldArea = styled.textarea<TextFieldAreaProps>`
  ::-ms-clear { display: none; }

  white-space: pre-wrap;
  word-break: break-word;

  ${getFontSize('medium')}
  color: ${themed(({ theme, disabled }) => (disabled ? theme.text3 : theme.text2))};
  border-radius: 0;
  font-family: inherit;
  padding: ${props => (props.labelShown ? `${distance.large} ${distance.medium} ${distance.small}` : distance.medium)};
  box-sizing: border-box;
  box-shadow: none;
  resize: ${props => props.resizable || 'none'};
  margin: 0;
  width: 100%;
  cursor: ${props => (props.disabled ? 'not-allowed' : 'auto')};
  border: none;
  border-bottom: 1px solid
    ${themed(({ theme, border }) => (border === TextFieldBorderType.error ? purpleRed : theme.ui4))};
  background: ${themed(({ theme }) => theme.ui2)};

  &::placeholder {
    color: ${themed(({ theme, disabled }) => (disabled ? theme.text3 : theme.text2))};
    opacity: ${props => (props.labelShown ? '0' : '1')};
    transition: inherit;
  }

  &:focus {
    outline: none;

    &::placeholder {
      opacity: 1;
    }
  }
`;

const TextFieldElement = styled.div`
  ${getFontSize('medium')}

  display: flex;
  align-items: center;
  padding: ${distance.medium};
  color: ${themed(({ theme }) => theme.text2)};
  background: ${themed(({ theme }) => theme.ui3)};
`;

class TextFieldInt extends React.Component<TextFieldProps & FormContextProps, TextFieldState> {
  private _element: TextFieldInputNodeType | null;

  constructor(props: TextFieldProps) {
    super(props);
    this.state = {
      focused: false,
      reveal: false,
      controlled: props.value !== undefined,
      value: props.value || props.defaultValue || '',
      error: props.error,
    };
  }

  componentWillReceiveProps({ value = '', error }: TextFieldProps) {
    if (this.state.controlled) {
      this.setState({ value });
    }
    this.setState({ error });
  }

  componentDidMount() {
    this.changeHeight();
    const { form } = this.props;
    const { controlled } = this.state;

    if (!controlled && form) {
      form.subscribe(this);
    }
  }

  componentWillUnmount() {
    const { form } = this.props;
    const { controlled } = this.state;

    if (!controlled && form) {
      form.unsubscribe(this);
    }
  }

  private changeValue = (e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    this.updateValue(e.target.value, e);
  };

  private updateValue(value: string, e?: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) {
    const { onChange, name = '', form } = this.props;

    if (!this.state.controlled) {
      if (form) {
        form.change({
          name,
          value,
        });
      } else {
        this.setState({
          value,
        });
      }
    }

    if (typeof onChange === 'function') {
      onChange({
        value,
        originalEvent: e,
      });
    }
  }

  private changeHeight = () => {
    const { resizable } = this.props;
    const element = this._element;

    if (resizable === 'auto' && element) {
      element.style.height = '1px';
      element.style.height = `${element.scrollHeight}px`;
    }
  };

  private setTextFieldNode = (node: TextFieldInputNodeType | null) => {
    this._element = node;
    const { inputRef } = this.props;
    if (typeof inputRef === 'function') {
      inputRef(node);
    }
  };

  private setFocus = () => {
    const { onFocus } = this.props;

    this.setState(
      {
        focused: true,
      },
      onFocus,
    );
  };

  private unsetFocus = () => {
    const { onBlur } = this.props;

    this.setState(
      {
        focused: false,
      },
      onBlur,
    );
  };

  private handleReset = () => {
    const { onClear } = this.props;
    this.updateValue('');

    if (typeof onClear === 'function') {
      onClear();
    }
  };

  render() {
    const {
      theme,
      borderless,
      prefix,
      suffix,
      multiline,
      resizable = false,
      disabled,
      placeholder,
      info,
      label,
      clearable,
      onClear,
      icon,
      onChange: _0,
      defaultValue: _1,
      value: _2,
      form: _3,
      onFocus: _4,
      onBlur: _5,
      inputRef: _6,
      onInput: _7,
      ...rest
    } = this.props;
    const { focused, value, error } = this.state;
    const rows = typeof multiline === 'number' ? multiline : undefined;
    const border = getTextFieldBorderType(borderless, !!error, focused);
    const hasValue = !!value;
    const input =
      rows !== undefined || multiline ? (
        <TextFieldArea
          ref={this.setTextFieldNode}
          onInput={this.changeHeight}
          border={border}
          rows={rows}
          resizable={typeof resizable === 'string' ? resizable : resizable ? 'auto' : undefined}
          disabled={disabled}
          labelShown={label !== undefined}
          placeholder={placeholder}
          onFocus={this.setFocus}
          onBlur={this.unsetFocus}
          value={value}
          onChange={this.changeValue}
          {...rest}
        />
      ) : (
        <StyledInput
          ref={el => this.setTextFieldNode(el as TextFieldInputNodeType)}
          theme={theme}
          disabled={disabled}
          labelShown={label !== undefined}
          placeholder={placeholder}
          onFocus={this.setFocus}
          onBlur={this.unsetFocus}
          value={value}
          onChange={this.changeValue}
          {...rest}
        />
      );
    const textFieldWrapper = (
      <StyledInputRow
        error={!!error}
        focused={focused}
        hasValue={hasValue}
        label={label}
        placeholder={placeholder}
        multiline={rows !== undefined || multiline !== undefined}>
        {input}
      </StyledInputRow>
    );

    return (
      <TextFieldContainer>
        {rows !== undefined || multiline ? (
          textFieldWrapper
        ) : (
          <TextFieldWrapper>
            {prefix && <TextFieldElement>{prefix}</TextFieldElement>}
            <StyledInputBox border={border} disabled={disabled} focused={focused} hasValue={hasValue}>
              {textFieldWrapper}
              <InputIcon
                disabled={disabled}
                defaultIcon={icon}
                theme={theme}
                error={error}
                hasValue={hasValue}
                clearable={clearable}
                onClear={this.handleReset}
                onClick={() => this._element && this._element.focus()}
              />
            </StyledInputBox>
            {suffix && <TextFieldElement>{suffix}</TextFieldElement>}
          </TextFieldWrapper>
        )}
        {showInputInfo(error, info)}
      </TextFieldContainer>
    );
  }
}

/**
 * A text field for custom user input.
 */
export const TextField = withFormContext(TextFieldInt);
TextField.displayName = 'TextField';
