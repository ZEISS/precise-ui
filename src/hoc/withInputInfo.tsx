import * as React from 'react';
import { InputInfo, InputInfoProps } from '../components/InputInfo';
import { InputError, InputErrorProps } from '../components/InputError';

type Props = {
  error?: React.ReactChild;
  info?: React.ReactChild;
};

export const withInputInfo = ({ error, info }: Props) => {
  const WithInfo = ({ info, ...props }: Pick<Props, 'info'>) => {
    return typeof info === 'string' ? (
      <InputInfo left="medium" {...props}>
        {info}
      </InputInfo>
    ) : (
      <>{info}</>
    );
  };
  const WithError = ({ error, ...props }: Pick<Props, 'error'>) => {
    return typeof error === 'string' ? (
      <InputError left="medium" {...props}>
        {error}
      </InputError>
    ) : (
      <>{error}</>
    );
  };

  if (error) return (props: InputErrorProps) => <WithError error={error} {...props} />;
  if (info) return (props: InputInfoProps) => <WithInfo info={info} {...props} />;

  return () => <></>;
};
