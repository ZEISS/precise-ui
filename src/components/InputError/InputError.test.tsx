import * as React from 'react';
import * as enzyme from 'enzyme';
import { InputError } from './';
import { TextField } from '../TextField';
import { withValidator } from '../../hoc/withValidator';

describe('<InputError />', () => {
  it('should render <InputError> that has an error message', () => {
    const wrapper = enzyme.shallow(<InputError>I am wrong</InputError>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should not render an <InputError> from jsx', () => {
    const wrapper = enzyme.shallow(<InputError />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should not render an <InputError> from direct API without any children', () => {
    const wrapper = enzyme.shallow(React.createElement(InputError));
    expect(wrapper).toMatchSnapshot();
  });

  it('should not be shown with the withValidator hoc if not changed', () => {
    const RequiredTextField = withValidator(TextField, ({ value }) => !value && 'Input is required');
    const wrapper = enzyme.mount(<RequiredTextField />);
    expect(wrapper.state('error')).toEqual(undefined);
  });

  it('should be shown with the withValidator hoc if changed', () => {
    const RequiredTextField = withValidator(TextField, ({ value }) => !value && 'Input is required');
    const wrapper = enzyme.mount(<RequiredTextField />);
    (wrapper.instance() as any).validate({
      value: '',
    });
    expect(wrapper.state('error')).not.toEqual('');
  });
});
