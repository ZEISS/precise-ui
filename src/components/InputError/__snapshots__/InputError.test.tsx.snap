// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputError /> should not render an <InputError> from direct API without any children 1`] = `""`;

exports[`<InputError /> should not render an <InputError> from jsx 1`] = `""`;

exports[`<InputError /> should render <InputError> that has an error message 1`] = `
<styled.div>
  I am wrong
</styled.div>
`;
